lib/spack/spack/spec.py

```diff
@@ -560,43 +560,47 @@ def __repr__(self):
 
 @key_ordering
 class DependencySpec(object):
-    """Dependencies can be one (or more) of several types:
+    """DependencySpecs connect two nodes in the DAG, and contain deptypes.
+
+    Dependencies can be one (or more) of several types:
 
     - build: needs to be in the PATH at build time.
     - link: is linked to and added to compiler flags.
     - run: needs to be in the PATH for the package to run.
 
     Fields:
-    - spec: the spack.spec.Spec description of a dependency.
-    - deptypes: strings representing the type of dependency this is.
+    - spec: Spec depended on by parent.
+    - parent: Spec that depends on `spec`.
+    - deptypes: list of strings, representing dependency relationships.
     """
 
-    def __init__(self, spec, deptypes, default_deptypes=False):
+    def __init__(self, parent, spec, deptypes):
+        self.parent = parent
         self.spec = spec
-        self.deptypes = deptypes
-        self.default_deptypes = default_deptypes
+        self.deptypes = tuple(sorted(set(deptypes)))
 
     def update_deptypes(self, deptypes):
-        if self.default_deptypes:
-            self.deptypes = deptypes
-            self.default_deptypes = False
-            return True
-        return False
-
-    def _cmp_key(self):
-        return self.spec
+        deptypes = tuple(sorted(set(deptypes)))
+        changed = self.deptypes != deptypes
+        self.deptypes = deptypes
+        return changed
 
     def copy(self):
```

Comment:
TBH I didn't know about `__copy__` and `__deepcopy__`.  But that sounds good to me!  There is one complication.  

If you notice, `Spec` has a `_dup` method, which does a copy in place.  The `copy()` method currently uses that.  It is used  when you call `.concretize()` on `Spec('mpi')`, you want it to become `mpich` or `openmpi` or whatever.

I have thought for a while that an immutable implementation of `Spec` might be a better way to go, but it would be a major rework, and I have no evidence to quantify whether it would be a win performance-wise or not.  On the one hand, the new concretizer will be iterating through a large space of `Specs`, and a completely immutable representation may hurt when we have to search a large space.  On the other hand, I think it would simplify the implementation of the concretizer and `Spec` hashing, and the protocol in #1821 and #1875.  So I guess my answer is let's open that can of worms when we get to it, which might be pretty soon here.